### DISCUSSION TITLE
Add nodeSelectors to manifestgen/install

### DISCUSSION THREAD

**Discussion by philnichol** [fluxcd/flux2#2422] [Proposals] [answered]
Since tolerations are supported when installing flux to a kubernetes cluster, additional custom nodeSelectors would be go hand-in-hand to allow better control of where the flux components reside.
The main purpose of this is to raise a follow-up in the terraform provider.

Raised this draft PR, let me know if it's the right direction?  https://github.com/fluxcd/flux2/pull/2423

If not happy for anyone else to raise a proper solution to this.

**Comment by kingdonb** [comment, marked as answer]
I didn't know that Flux can set tolerations at install time:

```
  flux install --toleration-keys=node.kubernetes.io/dedicated-to-flux
      --toleration-keys strings    list of toleration keys used to schedule the components pods onto nodes with matching taints
```

It makes sense to me that if we're going to support tolerations we also should support nodeSelectors, but I saw that we published this cheatsheet a few days ago which addresses tolerations and nodeSelector in the recommended way, with a patch:

https://fluxcd.io/docs/cheatsheets/bootstrap/#node-affinity-and-tolerations

I'm not sure how helpful that is for Terraform users or not, but maybe this solves your issue.

There might be a principled reason why we have chosen to support only tolerations and not nodeSelectors in the CLI (because `flux install` without setting toleration will fail on a cluster with only control plane nodes, you need the toleration to ensure the pods will launch, but you can set nodeSelectors later in a patch... that might be what we were going for there? Not sure exactly.)

So there is a workaround for nodeSelectors but without the tolerations flag there is no workaround for those users, (besides untainting a node, which might not be an option at all.)

Also I had a look at your PR and noticed you had an issue with timezones in the tests, then I found this:

* https://github.com/fluxcd/flux2/pull/2441

which looks like it addresses those failures 👍 

I'm not sure if manifestgen options surface automatically in the CLI, or if there is more work needed to merge your PR. Thanks for contributing this and opening a discussion about it. Let us know if you're blocked and need help getting this issue solved!

**Reply by philnichol** [reply]
Hi @kingdonb , thanks for the feedback and for thanks maintaining this awesome project!
 
I don't think they will automatically surface in the CLI, I believe this would only allow us to render nodeSelectors into the final yaml file. The manifestgen package is what the TF provider uses to set up flux install/sync.

Makes sense what you've said from a tolerations standpoint, and thanks for sharing the link with the patch cheatsheet. Will pass this on to the team and looks like it will sort us out 🙂 .

I understand that the patch will do the trick, but IMHO this extra flexibility would be nice to have as a proper flag rather than having to run a separate command.  We have a terraform module that spins up a cluster and bootstraps flux (which then goes on to install standard addons, etc), and we really like the idea of being able to use `flux_install` and `flux_sync` resources curated by the maintainers rather than having to add stuff which may or may not always work as expected.

Either way, you've provided a good solution, so happy to mark as answered, but if there's any appetite to consider the PR or adding further configuration options to manifestgen please let me know.
